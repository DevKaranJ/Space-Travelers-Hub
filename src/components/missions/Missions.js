import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions, joinMission, leaveMission } from '../../actions/missions/missions';

function Missions() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const error = useSelector((state) => state.missions.error);

  const handleJoinMission = (missionId) => {
    dispatch(joinMission(missionId));
  };

  const handleLeaveMission = (missionId) => {
    dispatch(leaveMission(missionId));
  };

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  if (error) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <div>
      <h2>Missions</h2>
      <ul>
        {missions.map((mission) => (
          <li key={mission.mission_id}>
            <h3>{mission.mission_name}</h3>
            <p>{mission.description}</p>
            {mission.reserved ? (
              <div>
                <span>Active Member</span>
                <button type="button" onClick={() => handleLeaveMission(mission.mission_id)}>Leave Mission</button>
              </div>
            ) : (
              <div>
                <span>NOT A MEMBER</span>
                <button type="button" onClick={() => handleJoinMission(mission.mission_id)}>Join Mission</button>
              </div>
            )}
          </li>
        ))}
      </ul>
    </div>
  );
}

export default Missions;
