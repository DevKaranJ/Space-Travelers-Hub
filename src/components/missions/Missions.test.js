import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Missions from './Missions';

const mockStore = configureStore([thunk]);

describe('Missions Component', () => {
  it('renders Missions component with sample data', async () => {
    const initialState = {
      missions: {
        missions: [
          {
            mission_id: 1,
            mission_name: 'Mission 1',
            description: 'Mission description 1',
            reserved: true,
          },
          {
            mission_id: 2,
            mission_name: 'Mission 2',
            description: 'Mission description 2',
            reserved: false,
          },
        ],
        error: null,
      },
    };

    const store = mockStore(initialState);

    const { container } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
