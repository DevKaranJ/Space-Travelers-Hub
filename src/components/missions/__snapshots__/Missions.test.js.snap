// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component renders Missions component with sample data 1`] = `
<div>
  <div
    class="mission"
  >
    <table
      class="mission-table"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Mission 1
          </td>
          <td>
            Mission description 1
          </td>
          <td>
            <div
              class="active-status"
            >
              Active Member
            </div>
          </td>
          <td>
            <button
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Mission 2
          </td>
          <td>
            Mission description 2
          </td>
          <td>
            <div
              class="inactive-status"
            >
              NOT A MEMBER
            </div>
          </td>
          <td>
            <button
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
